instance/models/redis.py

```diff
@@ -0,0 +1,70 @@
+# -*- coding: utf-8 -*-
+#
+# OpenCraft -- tools to aid developing and hosting free software projects
+# Copyright (C) 2015-2021 OpenCraft <xavier@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Definition of Redis-specific models.
+"""
+import redis
+from django.db import models
+from django_extensions.db.models import TimeStampedModel
+
+from instance.models.utils import ValidateModelMixin
+
+
+class RedisUser(ValidateModelMixin, TimeStampedModel):
+    """
+    A model representing a single Redis user identified by ACLs.
+    """
+
+    # TODO: The user can have multiple passwords set by ACLs. Create signals to
+    # handle ACL setting on user creation, ACL update on password change, and
+    # ACL deletion on user deletion.
```

Comment:
I feel like a safer method would be to add a methods `create_redis_user`, `delete_redis_user`, `set_password`, which would include the required acl setup/teardown steps.  Then we don't need to rely on signals and all the logic is in one place.

Reply:
(Similar to the `provision_rabbitmq`, `deprovision_rabbitmq` methods)

Reply:
@gabor-boros actually, why do we need this model at all?  Can't it all be handled by the `RedisInstanceMixin`?  Or is there a possibility of multiple redis users defined per instance? :thinking: After reading the code of both models, it seems that we're splitting logic across two models.

Reply:
> Can't it all be handled by the RedisInstanceMixin?

I spent some time thinking about this. Having the user info in a separate model is not a bad idea since the `RedisUser` model encapsulates the user management logic (create/delete). If that would be explicitly handled by the `RedisInstanceMixin` -- not just calling the user creation but actually know how to create a user -- it could be a mess soon if we need to do more Redis user-related calls.

I'd say that the rule of thumb here could be
* `RedisUser` model should have all Redis user-related logic
* `RedisInstanceMixin` should have all provisioning logic
* `RedisServer` model should have all Redis logic that is **not** user-related

Based on the above, getting the Redis client should that is passed around should be handled by the `RedisServer` model as it is responsible for the connection parameters including credentials. What do you think about this?

Reply:
Ok, using signals makes sense if we're adding/removing RedisUsers in django.  But we're tied to one redis 'user' per openedxinstance right?  Then we don't need to worry about creating/deleting RedisUsers, because the provisiong/deprovisioning of redis users/acls will be managed by provisioning/archiving openedx instances.

> RedisUser model should have all Redis user-related logic

Just be careful that we aren't making unecessary database models simply as a way to namespace functions. ;)

Reply:
In the RedisInstanceMixin, we can have the `redis_user` and `redis_password`, and move the `delete_acl` and `create_acl` methods to RedisInstanceMixin.provision_redis/deprovision_redis.  Then we have half as many methods and models to maintain.

Reply:
For example if we did something like this:

```diff
diff --git a/instance/models/mixins/redis.py b/instance/models/mixins/redis.py
index 65cefaae..fd065197 100644
--- a/instance/models/mixins/redis.py
+++ b/instance/models/mixins/redis.py
@@ -38,16 +38,6 @@ from instance.models.redis_server import RedisServer
 
 # Functions ###################################################################
 
-def new_redis_user() -> RedisUser:
-    """
-    Return a new Redis user.
-    """
-    return RedisUser.objects.create(
-        username=get_random_string(length=32, allowed_chars=string.ascii_lowercase),
-        password=get_random_string(length=64),
-    )
-
-
 def select_random_redis_server():
     """
     Helper for the field default of `redis_server`.
@@ -73,14 +63,8 @@ class RedisInstanceMixin(models.Model):
         on_delete=models.PROTECT,
     )
 
-    redis_consumer_user = models.ForeignKey(
-        RedisUser,
-        related_name='consumer_instance',
-        on_delete=models.CASCADE,
-        blank=True,
-        null=True,
-        default=new_redis_user,
-    )
+    redis_username = models.CharField(max_length=32, unique=True, default=lambda:get_random_string(length=32, allowed_chars=string.ascii_lowercase))
+    redis_password = models.CharField(max_length=64, default=lambda:get_random_string(length=64))
 
     redis_provisioned = models.BooleanField(default=False)
 
@@ -119,11 +103,19 @@ class RedisInstanceMixin(models.Model):
         self.logger.info('Provisioning Redis started.')
 
         client = self._redis_client()
-        self.redis_consumer_user.create_acl(client)
+        client.acl_setuser(
+            username=self.redis_username,
+            passwords=[f"+{self.redis_password}"],
+            keys=[self.get_key_prefix()],
+            enabled=True
+        )
 
         self.redis_provisioned = True
         self.save()
 
+    def get_key_prefix(self) -> str:
+        return self.redis_username + _
+
     def deprovision_redis(self, ignore_errors=False):
         """
         Deletes the Redis vhost and users.
@@ -141,7 +133,7 @@ class RedisInstanceMixin(models.Model):
 
         try:
             client = self._redis_client()
-            self.redis_consumer_user.delete_acl(client)
+            client.acl_deluser(self.redis_username)
         except Exception as exc:
             self.logger.exception(
                 'Cannot delete redis user: %s. %s',
```

Reply:
@swalladge I had some time to think about this:

I did apply the suggestion related to `RedisUser` and committed the changes already. Regarding my previous statement about signals, I still believe that we could have signals, though I would leave that as an improvement idea.

I mean, implementing the above-mentioned signals would make sense, though I cannot remember for a single time when I had to manually change the credentials of a rabbit mq vhost, so I believe that won't be often for redis either.

How about creating a ticket in the backlog about this improvement idea to not forget it, but skip the implementation for that now?

Reply:
@gabor-boros looks good :+1:

bq. I mean, implementing the above-mentioned signals would make sense...

When would a signal be required?  If we needed to update the credentials, wouldn't the flow be:

- change credentials
- launch new appserver
  - which will run the redis_provision method, updating the acls

Reply:
@swalladge Actually the workflow should be:
1. change credentials in ocim
2. delete ACL in Redis
3. uncheck `redis_provisioned`
4. launch new app server

Without step 2, the Redis instance would be still accessible with old credentials. Again, I cannot remember a single time when we had to change credentials so it is really rare.

Reply:
@gabor-boros 

> Without step 2, the Redis instance would be still accessible with old credentials.

Are you sure about that?  According to [the docs for redis.acl_setuser](https://redis-py.readthedocs.io/en/stable/#redis.Redis.acl_setuser):

> Create or update the ACL for username. If the user already exists, the existing ACL is completely overwritten and replaced with the specified values.

So there shouldn't be old acls hanging around.

Reply:
Oh, then I misread something! Sorry about that. In this case we are fine here.

Reply:
Great!  I also had in my mind that we would need manual processing to delete old ACLs, but maybe it's just that the python library handles this for us, while the redis client doesn't.  CC @pomegranited in case I missed something.

Reply:
As this seems to be resolved, I'm resolving the thread too.